docs/dev/roadmap/code_structure/directory_reference.md

```diff
@@ -0,0 +1,128 @@
+`DelugeFirmware/src/deluge`
+
+*Main directory for firmware code*
+
+Important files:
+
+| File | Relevance
+|-|-
+| [`deluge.cpp`](class_outlines/main_c.md) | `deluge_main()` is called from `main.c` and starts all repeating tasks
+
+| Folder | Contents
+|-|-
+| drivers | mostly deals with managing hardware
+| dsp | digital signal processing, manages audio effects
+| [gui](#gui-folder) | graphical user interface, manages all user-facing behavior of the Deluge
+| hid | human interface device, manages basic functionality of knobs, buttons, pads, LEDs, display
+| io | input/output, manages debug-logging and incoming midi
+| memory | memory management
+| [model](#model-folder) | manages model_stack, the model Deluge uses to represent and pass around its internal state
+| modulation | manages internal modulating of parameters, including automation and midi
+| playback | manages playback and recording modes
+| processing | manages various processes: audio, CV and live audio engines, sound instruments, metronome, stem export
+| storage | manages using files from the SD card
+| testing | automated tests
+| util | collection of general utilities
+| version | deals with the code version
+
+## GUI Folder
+`DelugeFirmware/src/deluge/gui`
+
+*Graphical User Interface. This folder manages all user-facing behavior of the Deluge and is therefor the main focus for fixing bugs and developing new features*
+
+| Folder | Contents
+|-|-
+| colour | colour definitions
+| context_menu | the context menu pops up on top of the regular menu, depending on contextual actions
+| fonts | font definitions
+| l10n | list of strings used in the ui for both 7SEG and oled, can be used for localization
+| menu_item | definitions for the options of all menus
+| ui | manages the foundational layer of the ui including menus, browsers, keyboard views and saving/loading. `ui.cpp` serves as a base-class that the different views/modes of the Deluge are built on
+| [views](#views-folder) | manages the main views/modes of the Deluge, such as clip view and arranger view
+| waveform | manages the waveform view
+
+### UI Folder
+`DelugeFirmware/src/deluge/gui/ui/`
+
+*User Interface. Many low-level and base methods of the user interface are defined here, and reused in higher level code*
+
+| Folder | Contents
+|-|-
+| browser | `Browser` is a `QwertyUI` which handles browsing files on the SD card. It has sub-classes for browsing DX7-presets (`DxSyxBrowser`,) audio samples (`SampleBrowser`,) and slots for instrument clips and songs (`SlotBrowser`)
+| keyboard | manages the keyboard overlay
+| load | `LoadUI` is a `SlotBrowser` with sub-classes `LoadInstrumentPresetUI`, `LoadMidiDeviceDefinitionUI` and `LoadSongUI`
+| rename | `RenameUI` is a `QwertyUI` with sub-classes `RenameClipUI`, `RenameDrumUI`, `RenameMidiCCUI` and `RenameOutputUI`
+| save | `SaveUI` is a `SlotBrowser` with sub-classes `SaveInstrumentPresetUI`, `SaveKitRowUI`, `SaveMidiDeviceDefinitionUI` and `SaveSongUI`
```

Comment:
probably needs a note about the slot browser itself? None of us really know what it's supposed to do since it was broken before the open source release but we think it's for managing old style deluge song files named like song123a, and the 'slots' are the number and the letter

Reply:
How much is this aimed at being planning / consensus building for future, and how much descriptions of current state of affairs?

If the latter, I'm marginally concerned about offside (ie. not mixed with code) documentation going out of sync. In my experience current state of affairs is better described with high level comments, or README.md / DESIGN.md / SUBSYSTEM_NAME.md files in the same directories as the source files.

There's also risk of implying "this is the design for a good reason" when something is just "this is the way it's always been": there's a lot of things like that... _Documenting_ them is super valuable, as long as we take care not to accidentally enshrine them as explicit design decisions :)

For explicit decisions, consensus building and "future documentation" I love this - 🚀🔥

(I _really_ like when decision docs mention the goals and forces involved, so it's easier to figure out if there's reason to revisit or not, but depending on the case that can increase the workload to create and maintain the docs quite a lot.)